lib/improver/weights.py

```diff
@@ -30,123 +30,205 @@
-                    Default value is None
-
-        Returns:
-            weights : array of weights
-                      sum of all weights = 1.0
-
-    """
-    if not isinstance(num_of_weights, int) or num_of_weights <= 0:
-        msg = ('Number of weights must be integer > 0 '
-               'num = {0:s}'.format(str(num_of_weights)))
-        raise ValueError(msg)
-    # Special case num_of_weighs == 1 i.e. Scalar coordinate.
-    if num_of_weights == 1:
-        weights = np.array([1.0])
-        return weights
-    if not isinstance(y0val, float) or y0val <= 0.0:
-        msg = ('y0val must be a float > 0.0, '
-               'y0val = {0:s}'.format(str(y0val)))
-        raise ValueError(msg)
-    if ynval is not None:
-        if slope != 0.0:
-            msg = ('Relative end point weight or slope must be set'
-                   ' but not both.')
+class WeightsUtilities(object):
+    """ Utilities for Weight processing. """
+
+    def __init__(self):
+        """Initialise class."""
+        pass
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<WeightsUtilities>')
+        return result
+
+    @staticmethod
+    def normalise_weights(weights):
+        """Ensures all weights add up to one.
+
+            Args:
+                weights : array of weights.
+
+            Returns:
+                normalised_weights : array of weights
+                                     where sum = 1.0
+
+            Raises:
+                ValueError 1 : any negative weights are found in input.
+                ValueError 2 : sum of weights in the input is 0.
+        """
+        if weights.min() < 0.0:
+            msg = ('Weights must be positive. The weights have at least one '
+                   'value < 0.0: {}'.format(weights))
+            raise ValueError(msg)
+
+        sumval = weights.sum()
+        if sumval == 0:
+            msg = 'Sum of weights must be > 0.0'
             raise ValueError(msg)
-        else:
-            slope = (ynval - y0val)/(num_of_weights - 1.0)
 
-    weights_list = []
-    for tval in range(0, num_of_weights):
-        weights_list.append(slope*tval + y0val)
+        normalised_weights = weights / sumval
+        return normalised_weights
 
-    weights = normalise_weights(np.array(weights_list))
+    @staticmethod
+    def redistribute_weights(weights, forecast_present, method='evenly'):
+        """Redistribute weights if any of the forecasts are missing.
+
+            Args:
+                weights : (numpy.ndarray) of weights.
+                forecast_present : (numpy.ndarray) size of weights
+                                  with values set as
+                                  1.0 for present
+                                  0.0 for missing.
+                method : (string) method to redistribute weights
+                         Options are
+                         evenly - adding the weights from the
```

Comment:
I've changed it to line up with the block above. It may still be slightly annoying though?